content/docs/iac/using-pulumi/extending-pulumi/build-a-provider.md

```diff
@@ -0,0 +1,756 @@
+
+### Features of the `file` provider
+
+The `file` provider can be used to create and modify a local file. Here's an example of how to use it in a Pulumi program:
+
+```yaml
+resources:
+  managedFile:
+    type: file:File
+    properties:
+      path: ${pulumi.cwd}/managed.txt
+      content: |
+        An important piece of information
+```
+
+In this example, we create a new resource called `managedFile` of type `file:File`. We can specify the path to write it to, and the contents that should be in it. During an update, Pulumi will use the `file` provider to ensure the file exists and has the specified contents.
+
+Now let's create the `file` provider that implements this.
+
+### Set up the project
+
+Since the provider is a separate code project from your Pulumi programs, the first step is to create a new directory for the provider code.
+
+```sh
+$ mkdir file-provider
+$ cd file-provider
+```
+
+{{< chooser language "go" >}}
+
+{{% choosable language go %}}
+
+#### Create `go.mod` file
+
+The `go.mod` file defined the Go project. It sets up the name of the Go module, the runtime requirements, and the module dependencies. We need the Pulumi Provider SDK (aka `github.com/pulumi/pulumi-go-provider`) as well as the standard Pulumi SDK (aka `github.com/pulumi/pulumi/sdk/v3`).
+
+```go
+module example.com/file-provider
+
+go 1.24
+
+toolchain go1.24.0
+
+require (
+	github.com/pulumi/pulumi-go-provider v1.0.0
+	github.com/pulumi/pulumi/sdk/v3 v3.167.0
+)
+```
+
+Once that file is created, go ahead and download the dependencies using `go get`:
+
+```sh
+$ go get example.com/file-provider
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+#### Create `PulumiPlugin.yaml` file
+
+The only other file you'll need is the `PulumiPlugin.yaml` file. This tells Pulumi two things: that this code can be loaded as a provider plugin, and what runtime environment to use for that.
+
+{{< chooser language "go" >}}
+
+{{% choosable language go %}}
+
+```yaml
+runtime: go
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+### Implement the interface
+
+{{< chooser language "go" >}}
+
+{{% choosable language go %}}
+
+To implment the provider, first create a file called `main.go` with the following contents:
+
+```go
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/property"
+)
+
+func main() {
+	err := p.RunProvider("file", "0.1.0", provider())
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %s", err.Error())
+		os.Exit(1)
+	}
+}
+
+func provider() p.Provider {
+	return infer.Provider(infer.Options{
+		Resources: []infer.InferredResource{infer.Resource[*File]()},
+		ModuleMap: map[tokens.ModuleName]tokens.ModuleName{
+			"file": "index",
+		},
+	})
+}
```

Comment:
There's an updated syntax based on a provider builder, let's use that here. For example:
https://github.com/pulumi/pulumi-go-provider/pull/361/files#diff-3bc712aa88b1d046861bb4595afcc752e29a8664895d708194341cd3b6be5312R96

Reply:
updated!